### PR TITLE
refactor: add extra region paramater to functions on Array, etc. that return new Arrays etc. (issue #4962)

### PR DESCRIPTION
**Pull request by stephentetley**
This PR adds an "output" region parameter to the signatures of functions on mutables (Array, MutDeque, MutList, MutMap, MutSet) that return new mutables.

Exception - `iterator` and `enumerator` functions and the `Iterable` class left untouched. Firstly, I was having to do extra copying / casting or both to fit using a new region inside the interator. Secondly and a worse problem, changing the signature of the `Iterable` class was having some strange name resolution problems that I couldn't resolve e.g. the (non-mutable) `Set.toString` stopped compiling where it uses `enumerator` in the comprehension.

The naming scheme of the new region params in the function signatures probably needs some discussion. I've wanted to avoid calling the region var `r` as `r` is usually used as a type variable in the existing Array (or MutList...) type and this looks like it would be confusing. I've given the new regions indices so they are usually `r1` but sometimes the existing mutables already use indexed type vars, e.g:

~~~

pub def transpose(r3: Region[r3], a: Array[Array[a, r1], r2]): Array[Array[a, r3], r3] \ { Read(r1, r2), Write(r3) } = ...

~~~

Maybe using `rn` as a mnemonic for "r(egion) n(ew)" would be reasonable choice for the variable name, type var names could continue to use an increasing index:

~~~

pub def transpose(rn: Region[r3], a: Array[Array[a, r1], r2]): Array[Array[a, r3], r3] \ { Read(r1, r2), Write(r3) } = ...

~~~

### CODE REVIEW

**Comment by magnus-madsen** [comment]
> The naming scheme of the new region params in the function signatures probably needs some discussion. I've wanted to avoid calling the region var r as r is usually used as a type variable in the existing Array (or MutList...) type and this looks like it would be confusing. 

I think we will be moving towards `rh: Region[r]` for "region handle" (but I have not yet fully decided).

**Comment by magnus-madsen** [comment]
OK, I see the naming issue. I think we should sort that out in the future.

File: main/src/library/Array.flix [hunk 1]
```diff
@@ -105,7 +105,8 @@ namespace Array {
     ///
     /// Equivalent to the expression `a[b..e]`.
     ///
-    pub def slice(b: Int32, e: Int32, a: Array[a, r]): Array[a, r] \ { Read(r), Write(r) } = a[b..e]
+    pub def slice(_: Region[r1], b: Int32, e: Int32, a: Array[a, r]): Array[a, r1] \ { Read(r), Write(r1) } =
+        unsafe_cast a[b..e] as Array[a, r1] \ { Read(r), Write(r1) }
```
Comment (magnus-madsen) [inline main/src/library/Array.flix:109]: This will be refactored, but in a separate PR.

**Comment by magnus-madsen** [comment]
Looks good; only issue is the complicated naming.

File: main/src/library/Array.flix (hunk 1 above)
Comment (stephentetley) [reply main/src/library/Array.flix:109]: Thanks - am I reverting the code?

Comment (magnus-madsen) [reply main/src/library/Array.flix:109]: No its fine for now :)

Comment (stephentetley) [reply main/src/library/Array.flix:109]: Oh yeah - it's already been merged.

**Pull request merged**